datafusion/src/logical_plan/builder.rs

```diff
@@ -287,16 +287,125 @@ impl LogicalPlanBuilder {
                 .into_iter()
                 .zip(join_keys.1.into_iter())
                 .map(|(l, r)| {
-                    let mut swap = false;
                     let l = l.into();
-                    let left_key = l.clone().normalize(&self.plan).or_else(|_| {
-                        swap = true;
-                        l.normalize(right)
-                    });
-                    if swap {
-                        (r.into().normalize(&self.plan), left_key)
-                    } else {
-                        (left_key, r.into().normalize(right))
+                    let r = r.into();
+                    let lr = l.relation.clone();
+                    let rr = r.relation.clone();
+
+                    match (lr, rr) {
+                        (Some(lr), Some(rr)) => {
+                            let l_is_left =
+                                self.plan.all_schemas().iter().any(|schema| {
+                                    schema.fields().iter().any(|field| {
```

Comment:
It would be cool if we could avoid the repetition here

Reply:
Agreed. I will see what we can do. I was trying to avoid heavy work for non-qualified tables but will revisit.